docs/how-it-works/internals.md

```diff
@@ -0,0 +1,120 @@
+# Internal Architecture
+
+> Last edited April 2024.
+
+This document will outline Lonboard's internal architecture, and hopefully should be useful for potential contributors.
+
+## Pydeck
+
+Before we consider Lonboard, we should briefly cover how Pydeck manages data transit.
```

Comment:
> how Pydeck manages data transit.

Definitely agree that serializing a Python dataframe to a string to pass via pydeck doesn't scale. Thus the easiest way to work with pydeck is via a URL in the data prop.

However Pydeck does also support a [Binary Data Transfer](https://deckgl.readthedocs.io/en/latest/binary_transfer.html) which uses the same idea as lonboard to transfer the binary payload via websockets. 

What is missing from pydeck (and arguably deck.gl by extension) is a way render the binary data on the other end. However, you have implemented a collection of [GeoArrow-compatible layers
](https://github.com/geoarrow/deck.gl-layers), is there any reason these couldn't be used via pydeck?

I did a quick test to compare pydeck vs lonboard rendering 3M points and while lonboard is faster, it certainly isn't orders of magnitude:

_videos trimmed to show time from notebook cell refresh to first paint_

https://github.com/developmentseed/lonboard/assets/453755/ebe7abb0-a636-4a39-8d34-39b04e351c62

https://github.com/developmentseed/lonboard/assets/453755/7931c874-c02a-454f-803a-9c62d3127f2b

Reply:
> However Pydeck does also support a [Binary Data Transfer](https://deckgl.readthedocs.io/en/latest/binary_transfer.html) which uses the same idea as lonboard to transfer the binary payload via websockets.

Yes, but with a few caveats:

- Pydeck supports only numeric binary arrays. So no strings nor nested structs or lists. This _includes_ any attribute columns, as far as I can tell. So if you want a tooltip with a string or struct column, you can't use binary transport.

    This also means it can't be used with any layer types other than points. Loaders.gl does have a binary data representation for other geometry types, but you'd have to manage conversion from Python geometries to that representation yourself. Which essentially re-implements GeoArrow.
- It's opt-in, and the user needs to know not only that the `use_binary_transport` param exists, but also its caveats, and which of their datasets can take advantage of the option.
- Pydeck's binary transport isn't compressed at all. I assume you tested in a local Python environment, where Python is on the same machine as the browser. Some of Devseed's clients work on remote JupyterLab (JupyterHub) environments, and so effective compression is a must. For large data sizes, gzip can be unacceptably slow (both to compress and to decompress, in my opinion). ZSTD is the best compression available for this type of data, but then you need a Wasm dependency on the frontend.
- Lonboard automatically manages data chunking. It looks like pydeck serializes columns individually but does not do any chunking. So for 3 million points it sends a single 48MB buffer for the `getPosition` accessor, which from a remote server can easily crash the websocket connection.

> What is missing from pydeck (and arguably deck.gl by extension) is a way render the binary data on the other end. However, you have implemented a collection of [GeoArrow-compatible layers ](https://github.com/geoarrow/deck.gl-layers), is there any reason these couldn't be used via pydeck?

No reason it couldn't be used, but you'd need to create Arrow tables. And pydeck looks like it's already using [binary data rendering](https://github.com/visgl/deck.gl/blob/33f8d2dd4ab98963f25c6e6914107491d89de143/bindings/pydeck/pydeck/data_utils/binary_transfer.py#L46-L49). It's unclear what pydeck would gain from using these layers except for non-point support. But then you'd need to either use Arrow over the network or manage your own buffer transport for nested types.